src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const arrayOfStates = [];
+  const stateCopy = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        const arrayToRemove = action.keysToRemove;
+
+        for (const key of arrayToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      case 'clear':
+        const arrayToDelete = Object.keys(stateCopy);
+
+        for (const key of arrayToDelete) {
+          delete stateCopy[key];
+        }
+        break;
+
+      default:
+        throw new Error('This action is not possible.');
+    }
+
+    arrayOfStates.push(Object.assign({}, stateCopy));
```

Comment:
```suggestion
    arrayOfStates.push({ ...stateCopy });
```
You can also use spread syntax to make a copy